### ISSUE TITLE
KD approach 

### ISSUE THREAD

**Issue by AnhLD2610** [songmzhang/DSKD#24]
i found that in dskd_cma_tinyllama.sh you using forward_kl instead of adaptive_kl as the best results in paper should i change the KD_OBJ from "forward_kl" to "adaptive_kl".

**Comment by songmzhang** [comment]
Yes, forward_kl is just a default kd objective in our scripts. You can change it to other divergence and see which performs better. The best kd objective may be different for different models and datasets.

**Comment by AnhLD2610** [comment]
![image](https://github.com/user-attachments/assets/bfd373e0-edba-4e49-9635-88e5e4c4b2b6)
both the dist_func and compute_forward_kl_divergence should be the same KL right

**Comment by songmzhang** [comment]
> ![image](https://private-user-images.githubusercontent.com/95870820/391830838-bfd373e0-edba-4e49-9635-88e5e4c4b2b6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzMyMDgyNzYsIm5iZiI6MTczMzIwNzk3NiwicGF0aCI6Ii85NTg3MDgyMC8zOTE4MzA4MzgtYmZkMzczZTAtZWRiYS00ZTQ5LTk2MzUtODhlNWU0YzRiMmI2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjAzVDA2MzkzNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZjYTQ4ZmNiNDI3YTlmMDA4Zjk0NWI2ZTg5NTdlYWU3ZWEyYjhhNDE5ZGNiYzY5YTY2MWJjMWQ4MmY3NmMxNzgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.XBKFA77JFRp3oeBDMgjgnVkMoky5UlMjiRKeXekcPA0) both the dist_func and compute_forward_kl_divergence should be the same KL right

No. Please note that here we fix the dist_func as forward_kl for `s2t_kd_loss (kd in the teacher space)` since we found it usually performed better than other divergence. And we can adjust the dist_func for `t2s_kd_loss (kd in the student space)` in the training scripts.

**Comment by AnhLD2610** [comment]
![image](https://github.com/user-attachments/assets/026eb53c-259a-4762-9ed9-595fba778134)
but in the paper you report that akl gave the best result that make me confuse. So the s2t_kd_loss (kd in the teacher space) is forward and t2s_kd_loss (kd in the student space) is adaptive_kl right?

**Comment by songmzhang** [comment]
Yes, this presentation is indeed a little confusing. As you said, for any DSKD/DSKD-CMA, dist_func is only adjusted for t2s_kd_loss, and s2t_kd_loss uses fixed forward_kl.

**Comment by AnhLD2610** [comment]
Thank you for the reply

**Comment by AnhLD2610** [comment]
i found that training with akl has worse performance than normal kl. Am i missing something? As your paper report that AKL has the best performance. From mistral to tinyllama

**Comment by songmzhang** [comment]
AKL does not necessarily lead to better performance. As we found later, the performance will be affected by models, data and even GPUs. But empirically, we suggest you try forward KL, reverse KL and AKL and select the best one from them.

**Comment by AnhLD2610** [comment]
![image](https://github.com/user-attachments/assets/d4f4c34b-14f5-4d02-979b-63e7ce1088d3)
i have try to reproduce your result on tinyllama and teacher is mistral but the result in S-NI and UnNI lower than the paper report, am i missing something, i run exactly the code you provide.

**Comment by songmzhang** [comment]
> ![image](https://private-user-images.githubusercontent.com/95870820/393499652-d4f4c34b-14f5-4d02-979b-63e7ce1088d3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM1Njc2MDEsIm5iZiI6MTczMzU2NzMwMSwicGF0aCI6Ii85NTg3MDgyMC8zOTM0OTk2NTItZDRmNGMzNGItMTRmNS00ZDAyLTk3OWItNjNlN2NlMTA4OGQzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjA3VDEwMjgyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQ1OGIzNzlkZjQ3M2IyMDJlMzZmZDJkNDkyM2Y4OTBmNGE4YWZlMjFlYmU5YWE2ODc1ZjE4MTFhYjgyNGQ2YmEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.ol2N5mkEGXQ1-ESHON8fvFxa6X6KqWHiaxbxr64z6qg) i have try to reproduce your result on tinyllama and teacher is mistral but the result in S-NI and UnNI lower than the paper report, am i missing something, i run exactly the code you provide.

May I ask which kd_objective you use?

**Comment by AnhLD2610** [comment]
i use the forward KL, but the adaptive KL give the same result, not change much.

**Comment by songmzhang** [comment]
These results look normal for forward KL. 

For AKL, you can try to modify the `.le()` to `.lt()` in the following line:
https://github.com/songmzhang/DSKD/blob/57d2290ce2d448be6293d97a82c6608addf33cdb/code/criterions/various_divergence.py#L172

Since AKL had not been open-sourced by the authors when we released our code, `.lt()` was our original re-implementation, which got the results in our paper. I'm not sure how this simple modification will impact the performance. You can try and see whether the original one yields better performance. Also, I noticed that the official implementation of [AKL](https://github.com/wutaiqiang/LLM_KD_AKL/blob/main/AKL.py) has now been released, which is also suggested to try.

**Comment by AnhLD2610** [comment]
I think the results may vary depending on the GPU. Can I ask for the code you used to get the results of SeqKD?

**Comment by songmzhang** [comment]
> I think the results may vary depending on the GPU. Can I ask for the code you used to get the results of SeqKD?

We did not use additional code for SeqKD, but just got eval results of the teacher model on the training set and then ran student SFT on it.
Specifically, you can get the responses of the teacher model on the training set by modifying `eval_main.sh` like this:
```
#! /bin/bash
set -e
export CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7,8
MASTER_ADDR=localhost
MASTER_PORT=29500
NNODES=1
NODE_RANK=0
GPUS_PER_NODE=8

DISTRIBUTED_ARGS="--nproc_per_node $GPUS_PER_NODE \
                  --nnodes $NNODES \
                  --node_rank $NODE_RANK \
                  --master_addr $MASTER_ADDR \
                  --master_port $MASTER_PORT"

# model
BASE_PATH=path_to_dskd_project
CKPT_PATH=path_to_teacher_checkpoint
MODEL_TYPE="gpt2"    # gpt2, qwen, mistral, llama

# task
TASK="eval_main"

# data
DATA_NAME=dolly
DATA_DIR="${BASE_PATH}/data/${DATA_NAME}"
DATA_NUM=-1      # "-1" means evaluation on all data

# hp
EVAL_BATCH_SIZE=16     # depends on your GPU memory
SEED=1    

# runtime
SAVE_PATH=$(dirname ${CKPT_PATH})

OPTS=""
# model
OPTS+=" --base-path ${BASE_PATH}"
OPTS+=" --model-path ${CKPT_PATH}"
OPTS+=" --n-gpu ${GPUS_PER_NODE}"
OPTS+=" --model-type ${MODEL_TYPE}"
# task
OPTS+=" --task ${TASK}"
# data
OPTS+=" --data-dir ${DATA_DIR}"
OPTS+=" --data-names ${DATA_NAME}"
OPTS+=" --num-workers 0"
OPTS+=" --dev-num ${DATA_NUM}"
OPTS+=" --data-process-workers -1"
OPTS+=" --json-data"
# hp
OPTS+=" --eval-batch-size ${EVAL_BATCH_SIZE}"
OPTS+=" --max-length 512"
OPTS+=" --max-prompt-length 256"
# runtime
OPTS+=" --do-eval"
OPTS+=" --save-dir ${SAVE_PATH}"
OPTS+=" --seed ${SEED}"
# deepspeed
OPTS+=" --deepspeed"
OPTS+=" --deepspeed_config ${BASE_PATH}/configs/deepspeed/ds_config_bf16.json"
# gen
OPTS+=" --do-sample"
OPTS+=" --top-k 0"
OPTS+=" --top-p 1.0"
OPTS+=" --temperature 1.0"


export NCCL_DEBUG=""
export TOKENIZERS_PARALLELISM=false
export PYTHONIOENCODING=utf-8
export PYTHONPATH=${BASE_PATH}
CMD="torchrun ${DISTRIBUTED_ARGS} ${BASE_PATH}/code/evaluate.py ${OPTS}"
echo ${CMD}

${CMD}
```

And another modification is to change "valid" in L28 to "train":
https://github.com/songmzhang/DSKD/blob/57d2290ce2d448be6293d97a82c6608addf33cdb/code/evaluate_main.py#L23-L33

After running the script, you will find a file named `answers.jsonl` in the directory of the teacher checkpoint, which records the responses of the teacher corresponding to the training data. You can replace the `"output"` of the training data `train.jsonl` with the`"text"` in `answers.jsonl` to get the new training data, which can be used for student SFT (SeqKD).

We will complement the code of this part as soon as possible.

**Comment by AnhLD2610** [comment]
Thank you for your quick and helpful response!

**Comment by wutaiqiang** [comment]
> AKL does not necessarily lead to better performance. As we found later, the performance will be affected by models, data and even GPUs. But empirically, we suggest you try forward KL, reverse KL, and AKL and select the best one from them.

Indeed, I found the same issue and the results would be unstable even on different random seeds. btw, the dolly train set is like a domain specific fine-tuning dataset so the KD process is task-specific. I suggest evaluating the distilled models on fewer epochs, as **more** epochs would lead to **higher results on the dolly test and lower results on other test datasets**. In a nutshell, the training data for distillation is _not proper_. The ideal dataset to distill is the original pretraining data, which is not available and also not affordable. The elegant way is to perform distillation when training, just like LLaMA 3.2 1B and 3B, rather than post-compression.

**Comment by songmzhang** [comment]
> > AKL does not necessarily lead to better performance. As we found later, the performance will be affected by models, data and even GPUs. But empirically, we suggest you try forward KL, reverse KL, and AKL and select the best one from them.
> 
> Indeed, I found the same issue and the results would be unstable even on different random seeds. btw, the dolly train set is like a domain specific fine-tuning dataset so the KD process is task-specific. I suggest evaluating the distilled models on fewer epochs, as **more** epochs would lead to **higher results on the dolly test and lower results on other test datasets**. In a nutshell, the training data for distillation is _not proper_. The ideal dataset to distill is the original pretraining data, which is not available and also not affordable. The elegant way is to perform distillation when training, just like LLaMA 3.2 1B and 3B, rather than post-compression.

Agree with you. The effectiveness of KD highly relies on the training data for distillation. When we talk about "Knowledge" distillation, the best phase to inject knowledge from data or a teacher model to a model should be pre-training. However, even in continual pre-training, it requires high computational costs (often >50B tokens) to achieve significant improvement on general benchmarks like MMLU. By contrast, distillation in post-training is more like learning how to better respond to human instructions or how to finish a specific task better in an efficient way. In this phase, I would more like to regard it as a "student model improvement" method rather than a "teacher model compression" method. It brings little new knowledge but may help better elicit the original knowledge in the model (like the gains in SFT and RLHF).

**Comment by wutaiqiang** [comment]
Yes. The author of Minillm also finds such things which can be found in his recent paper "MiniPLM"

<img width="742" alt="image" src="https://github.com/user-attachments/assets/c2e491cf-0f95-4c0f-a646-4e7966cd5e7f" />

**Comment by AnhLD2610** [comment]
> These results look normal for forward KL.
> 
> For AKL, you can try to modify the `.le()` to `.lt()` in the following line:
> 
> https://github.com/songmzhang/DSKD/blob/57d2290ce2d448be6293d97a82c6608addf33cdb/code/criterions/various_divergence.py#L172
> 
> Since AKL had not been open-sourced by the authors when we released our code, `.lt()` was our original re-implementation, which got the results in our paper. I'm not sure how this simple modification will impact the performance. You can try and see whether the original one yields better performance. Also, I noticed that the official implementation of [AKL](https://github.com/wutaiqiang/LLM_KD_AKL/blob/main/AKL.py) has now been released, which is also suggested to try.

[adaptive.txt](https://github.com/user-attachments/files/18224104/adaptive.txt)
here is my train file as you suggest but it not change anything and the result still worse than forwardKL
[forwardkl.txt](https://github.com/user-attachments/files/18224114/forwardkl.txt)

**Comment by songmzhang** [comment]
> > These results look normal for forward KL.
> > For AKL, you can try to modify the `.le()` to `.lt()` in the following line:
> > https://github.com/songmzhang/DSKD/blob/57d2290ce2d448be6293d97a82c6608addf33cdb/code/criterions/various_divergence.py#L172
> > 
> > Since AKL had not been open-sourced by the authors when we released our code, `.lt()` was our original re-implementation, which got the results in our paper. I'm not sure how this simple modification will impact the performance. You can try and see whether the original one yields better performance. Also, I noticed that the official implementation of [AKL](https://github.com/wutaiqiang/LLM_KD_AKL/blob/main/AKL.py) has now been released, which is also suggested to try.
> 
> [adaptive.txt](https://github.com/user-attachments/files/18224104/adaptive.txt) here is my train file as you suggest but it not change anything and the result still worse than forwardKL [forwardkl.txt](https://github.com/user-attachments/files/18224114/forwardkl.txt)

Hi, I've checked the logs you provided and I did not see much difference with ours. I think the training works normally. The performance difference may be from the seeds, GPUs, and the teacher models. BTW, we found that the performance on S-NI and UnNI are highly sensitive to model changes since they are ood test sets (as shown in below). So the difference might not be that big as it shows.
<img width="660" alt="image" src="https://github.com/user-attachments/assets/fb51685d-2dd4-457b-91ce-c859bae8ae8e" />
<img width="663" alt="image" src="https://github.com/user-attachments/assets/cb769827-1c27-4411-bea8-9d69768444b4" />

**Comment by AnhLD2610** [comment]
Thank you for your meaningful response!

**Comment by AnhLD2610** [comment]
When reproducing the baseline for Mined and ULD, I obtained a result of around Rouge-L 25 from Mistral to TinyLLaMA. However, your paper reports only 23 and 20, respectively. Am I missing something?
